File: packages/flutter/test/state_resolution_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:bharat_choropleth/src/states.dart';
import 'package:flutter_test/flutter_test.dart';

/// The other half of the guard described in `lib/src/states.dart`.
///
/// `packages/js/test/state-resolution-cases.json` records every spelling the
/// JavaScript registry accepts and the id it resolves to. This replays all of
/// them against the Dart translation. The two registries cannot be diffed as
/// text, so they are held together by behaviour instead: if either drifts, this
/// fails, or its JavaScript counterpart does.
void main() {
  final fixture = jsonDecode(
    File('../js/test/state-resolution-cases.json').readAsStringSync(),
  ) as Map<String, dynamic>;
  final resolves = (fixture['resolves'] as Map<String, dynamic>).cast<String, String>();
  final doesNotResolve = (fixture['doesNotResolve'] as List<dynamic>).cast<String>();

  group('state registry parity with the JavaScript package', () {
    test('holds the same number of states', () {
      expect(kStates.length, fixture['states']);
    });

    test('resolves every recorded spelling to the same id', () {
      final mismatches = <String>[];
      resolves.forEach((spelling, expectedId) {
        final actual = resolveState(spelling)?.id;
        if (actual != expectedId) mismatches.add('"$spelling" -> $actual, expected $expectedId');
      });
      expect(mismatches, isEmpty, reason: 'Dart and JavaScript registries disagree:\n${mismatches.join('\n')}');
    });

    test('declines the spellings recorded as unresolvable', () {
      for (final input in doesNotResolve) {
        expect(resolveState(input), isNull, reason: '"$input" should not resolve');
      }
    });

    test('normalizes the way the JavaScript registry does', () {
      expect(normalizeStateKey('Tamil Nadu'), 'tamil-nadu');
      expect(normalizeStateKey('TAMIL-NADU'), 'tamil-nadu');
      expect(normalizeStateKey('tamil_nadu'), 'tamil-nadu');
      expect(normalizeStateKey('Jammu & Kashmir'), 'jammu-and-kashmir');
      expect(normalizeStateKey('  Goa  '), 'goa');
    });
  });
}

File: packages/js/test/state-resolution.test.ts
import { describe, expect, it } from "vitest";
import cases from "./state-resolution-cases.json";
import { STATES, resolveState } from "../src/states";

/**
 * The fixture is this registry's own output, recorded so the Flutter package can
 * be held to it — `packages/flutter/lib/src/states.dart` is a translation, and
 * nothing can diff Dart against TypeScript.
 *
 * This half asserts the recording is still accurate. Change the registry without
 * regenerating, and this fails with the instruction; the Dart test then fails too
 * until it matches. Regenerate with:
 *
 *     pnpm --filter bharat-choropleth-js build
 *     node scripts/generate-state-resolution-cases.mjs
 */
describe("state resolution cases (shared with the Flutter package)", () => {
  it("covers every state in the registry", () => {
    expect(cases.states).toBe(STATES.length);
  });

  it("still resolves exactly as recorded", () => {
    const actual = Object.fromEntries(
      Object.keys(cases.resolves).map((spelling) => [spelling, resolveState(spelling)?.id]),
    );
    expect(actual).toEqual(cases.resolves);
  });

  it("still declines the spellings recorded as unresolvable", () => {
    for (const input of cases.doesNotResolve) {
      expect(resolveState(input), `${JSON.stringify(input)} should not resolve`).toBeUndefined();
    }
  });
});
